pyc64/realemulator.py

```diff
@@ -52,10 +74,142 @@ def run_rom_code(self, reset):
+                self.load(destinationAddress,prog)  
+                cpu.pc=0xf5a9
+                return            
+            else:
+                final_path="drive{}/{}".format(fa, fname)
+                try:
+                    with open(final_path, "rb") as file:
+                        startAddr = struct.unpack("<H", file.read(2))[0]                                            
+                        prog=file.read()
+                        endAddress=self.load(startAddr,prog)
+                        print("\nLoading {:02X}:{:02X} {} Start Addr: {:02X} ... up to: $ {:02X}\n".format(
+                                       fa,sa,final_path, startAddr, endAddress))                        
+                        file.close()                
+                    # success
+                    cpu.pc=0xf5a9
+                except FileNotFoundError:
+                    print("ERR FILE NOT FOUND:", final_path)
+                    cpu.pc=0xf704 # 'file not found'           
+        else:
+            print("device not present (VERIFY command not supported)")
+            cpu.pc=0xf707
+
+    def load(self,startAddr,prog):
+        """
+         Load a program into memory.
+         Emulate Kernal code for prg loads
+        """
+        ram = self.screen.memory
+        cpu = self.real_cpu_running
+        endAddress= startAddr + len(prog)
+        ram[startAddr: endAddress] = prog
+        ram[0x90] = 0  # status OK                
+        ram[0xae]= endAddress & 0x00ff
+        high=(endAddress & 0xff00) >>8                    
+        ram[0xaf]= high
+        return endAddress        
+
+    def get_filename(self,fnaddr,fnlen,cpu):
+        fname=""
+        for i in range(0,fnlen):
+            fname=fname + chr(cpu.ByteAt(fnaddr+i)).lower()
+        if not fname.endswith(".prg") and fname!="$":
+            fname=fname+".prg"
+        return fname         
+
+    def make_dir_listing(self,deviceNumber,basicLoadAddress):
+        listing=[]
+        address = basicLoadAddress
+        def add_line(lineNumber:int, line: str):
+            nonlocal address
+            address = address + len(line)+3
+            listing.append(address & 0xff)
+            listing.append( (address & 0xff00) >>8 )
+            listing.append( lineNumber & 0xff)
+            listing.append( (lineNumber & 0xff00) >>8 )
+            data=bytearray(line,"utf-8")
+            for d in data:
+                listing.append(d)
+            listing.append(0)        
+        # For formatting see https://pyformat.info/#string_pad_align
+        add_line(0, "\u0012\"{:16.16}\" 00 2A".format("DRIVE"+str(deviceNumber)))
+        # scan directory and find out files.
+        # Then produce a "floppy disk drive"-like directory
+        total_blocks=0
+        for root, dirs, filenames in os.walk("./drive{}".format(deviceNumber)):
+            for fname in fnmatch.filter(filenames,"*.prg"):                
```

Comment:
.... why only *.prg?  I don't like the arbitrary restriction. Can you make it to just list all the files in the directory, regardless of the suffix?

Reply:
The code is not able to run bas file and I'd like to avoid showing them.
I will fix it in the branch.  :)

Reply:
I see, fair enough.  When you save something from basic it will become a .prg anyway.  So it's fine as it is really